File: backend/app/core/events/subscribers/session_broadcast.py
from __future__ import annotations

import logging

from app.core.events.envelope import ConductorEventEnvelope
from app.core.events.subscriber import SubscriberPhase
from app.core.session_broadcaster import SessionBroadcaster

logger = logging.getLogger(__name__)


class SessionBroadcastSubscriber:
    """Phase 2: broadcast session events to WebSocket subscribers via SessionBroadcaster."""

    name = "session_broadcast"
    phase = SubscriberPhase.BROADCAST

    def __init__(self, broadcaster: SessionBroadcaster) -> None:
        self._broadcaster = broadcaster

    async def handle(self, envelope: ConductorEventEnvelope) -> None:
        if envelope.session_id is None:
            return

        event_dict = {"type": envelope.event_type}

        if envelope.is_status_change:
            if envelope.stop_reason:
                event_dict["stop_reason"] = envelope.stop_reason
        else:
            if envelope.seq:
                event_dict["seq"] = envelope.seq
            if envelope.payload:
                event_dict.update(envelope.payload)

        await self._broadcaster.send(envelope.session_id, event_dict)

File: backend/app/joysafeter_orchestrator_rs/src/events/session_broadcast.rs
use std::sync::Arc;

use tokio::sync::broadcast;
use tracing::warn;

use super::envelope::EventEnvelope;
use crate::kernel::session_broadcaster::SessionBroadcaster;

/// SessionBroadcastSubscriber — BROADCAST phase.
///
/// Sends session-level events to WebSocket/SSE subscribers via SessionBroadcaster.
/// Mirrors the Python `SessionBroadcastSubscriber`.
pub struct SessionBroadcastSubscriber {
    broadcaster: SessionBroadcaster,
}

impl SessionBroadcastSubscriber {
    pub fn new(broadcaster: SessionBroadcaster) -> Self {
        Self { broadcaster }
    }

    /// Spawn as a background task listening on the event bus.
    pub fn spawn(
        self,
        mut rx: broadcast::Receiver<Arc<EventEnvelope>>,
    ) -> tokio::task::JoinHandle<()> {
        tokio::spawn(async move {
            loop {
                match rx.recv().await {
                    Ok(envelope) => {
                        self.handle(&envelope).await;
                    }
                    Err(broadcast::error::RecvError::Lagged(n)) => {
                        warn!("SessionBroadcastSubscriber lagged by {n} messages");
                    }
                    Err(broadcast::error::RecvError::Closed) => break,
                }
            }
        })
    }

    async fn handle(&self, envelope: &EventEnvelope) {
        // Runner/status envelopes must only reach the UI after they have been
        // persisted and assigned the canonical DB seq. Raw envelopes can carry
        // runner seq/no-seq values, which makes live ordering differ from the
        // refresh path that reads from DB.
        if !envelope.is_status_change || !envelope.db_persisted || envelope.session_seq.is_none() {
            return;
        }

        let mut event = envelope.payload.clone();
        if !event.is_object() {
            event = serde_json::json!({});
        }
        if let Some(obj) = event.as_object_mut() {
            obj.insert("type".to_string(), serde_json::json!(envelope.event_type));
            if let Some(id) = envelope.event_id {
                obj.insert("id".to_string(), serde_json::json!(id.to_public()));
            }
            if let Some(seq) = envelope.session_seq {
                obj.insert("seq".to_string(), serde_json::json!(seq));
            }
            if !obj.contains_key("stop_reason") {
                obj.insert(
                    "stop_reason".to_string(),
                    envelope.stop_reason.clone().unwrap_or_default(),
                );
            }
        }

        self.broadcaster.send(envelope.session_id, event).await;
    }
}
